Tensor construction: combine Resize+mutable_data - 1/4

jerryzh168 opened the pull request:
Summary:
Codemod generated with clangr shard mode, 25 files per diff,
motivation: https://github.com/pytorch/pytorch/pull/12407

Differential Revision: D13007246

smessmer commented on caffe2/operators/batch_gather_ops.h:
```diff
@@ -22,15 +22,14 @@ class BatchGatherOp final : public Operator<Context> {
     shape.insert(shape.end(), indices.sizes().begin(), indices.sizes().end());
     shape.insert(shape.end(), data.sizes().begin() + 2, data.sizes().end());
-    output->Resize(shape);
+    auto* output = Output(0, shape, at::dtype<float>());
```
The code below doesn't look like this is always float. It only uses float if data.IsType<float>(). It uses data.dtype(), maybe you should use that too?

smessmer commented on caffe2/operators/cross_entropy_op.cc:
```diff
@@ -78,13 +78,15 @@ bool SigmoidCrossEntropyWithLogitsOp<float, CPUContext>::RunOnDevice() {
-    out->Resize(dims);
+    std::vector<int64_t> dims_(
+        logits.sizes().begin(), logits.sizes().end() - 1);
+    dims = dims_;
```
This code seems overly complicated. Proposing some changes that make it more readable, and also faster (because it avoids the vector copy you have in line 87 and moves instead).

    std::vector<int64_t> dims;
    if (logits.dim() != 0) {
      dims = std::vector<int64_t>(
          logits.sizes().begin(), logits.sizes().end() - 1);
    }

smessmer commented on caffe2/operators/cross_entropy_op.cc:
```diff
@@ -160,13 +162,16 @@ bool WeightedSigmoidCrossEntropyWithLogitsOp<float, CPUContext>::RunOnDevice() {
-    out->Resize(dims);
+    std::vector<int64_t> dims_(
+        logits.sizes().begin(), logits.sizes().end() - 1);
+    dims = dims_;
```
same here

smessmer commented on caffe2/operators/elementwise_ops.h:
```diff
@@ -147,19 +147,19 @@ class BinaryElementwiseWithArgsOp final : public Operator<Context> {
-          &B,
+      CAFFE_ENFORCE(
+          !IsInputOutputAlias(1, 0),
           "In-place is allowed only with the first tensor when "
```
this wasn't changed in this diff, but this message is weird. if it is legacy broadcast, then it asserts that it is not inplace, saying "inplace is only allowed if legacy broadcast"? Is this a bug?

jerryzh168 replied:
looks like this is asserting we can only allow Output(0) being alias of Input(0), this is checking Output(0) is not an alias of Input(1). cc @BIT-silence

smessmer commented on caffe2/operators/elementwise_ops.h:
```diff
@@ -176,17 +176,19 @@ class BinaryElementwiseWithArgsOp final : public Operator<Context> {
-        CAFFE_ENFORCE_EQ(C_dims, B_dims);
-      } else {
-        C->Resize(C_dims);
+      std::copy(
```
m we should probably change the return type of ComputeBinaryBroadcastForwardDims to vector<int64_t>, so this weird copy isn't necessary. Probably shouldn't be done in this diff though.

jerryzh168 replied:
yeah, there are some cases in "caffe2/utils/math.h" as well

jerryzh168 committed:
Tensor construction: combine Resize+mutable_data - 1/4 (#13853)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13853

Codemod generated with clangr shard mode, 25 files per diff,
motivation: https://github.com/pytorch/pytorch/pull/12407

Reviewed By: smessmer

Differential Revision: D13007246

fbshipit-source-id: 229ad70035f30f7dff3c7fe26bdec99049e2abba
